### PR TITLE
Rmc 145b.1 fulfilment event

### PR DESCRIPTION
**Pull request by hbrace**
# Motivation and Context

We need a way sending SMS messages with a UAC from GOV NOTIFY.

 # What has changed

New service to turn a fulfilment request for a UAC into a request to GOV NOTIFY api.

 # How to test?

Run the acceptance tests

 # Links

https://trello.com/c/yX4y7G8z/944-rmc-145b-implement-uac-fulfilment-event-part-a-13-part-b-5
 # Screenshots (if appropriate):

### CODE REVIEW

**Comment by hbrace** [comment]
[RMC-145(B.1) Implement UAC fulfilment event (13)](https://trello.com/c/yX4y7G8z/944-rmc-145b-implement-uac-fulfilment-event-part-a-13-part-b-5)

File: src/test/java/uk/gov/ons/census/notifyprocessor/dto/NotifyApiResponse.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
+package uk.gov.ons.census.notifyprocessor.dto;
+
+import java.util.UUID;
+import lombok.Data;
+
+@Data
+public class NotifyApiResponse {
+  UUID id;
```
Comment (NickGrantONS) [inline src/test/java/uk/gov/ons/census/notifyprocessor/dto/NotifyApiResponse.java]: can make these members private if you want

File: src/test/java/uk/gov/ons/census/notifyprocessor/messaging/FulfilmentRequestReceiverIT.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+@ContextConfiguration
+@ActiveProfiles("test")
+@SpringBootTest
+@DirtiesContext(classMode = DirtiesContext.ClassMode.AFTER_EACH_TEST_METHOD)
+@RunWith(SpringJUnit4ClassRunner.class)
+public class FulfilmentRequestReceiverIT {
+
+  @Autowired private RabbitQueueHelper rabbitQueueHelper;
+
+  @Rule public WireMockRule wireMockCaseApi = new WireMockRule(wireMockConfig().port(8089));
```
Comment (NickGrantONS) [inline src/test/java/uk/gov/ons/census/notifyprocessor/messaging/FulfilmentRequestReceiverIT.java]: it's both case API and Notify API, so name might be misleading

File: src/test/java/uk/gov/ons/census/notifyprocessor/utilities/RabbitQueueHelper.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    SimpleMessageListenerContainer container =
+        new SimpleMessageListenerContainer(connectionFactory);
+    container.setMessageListener(messageListener);
+    container.setQueueNames(queueName);
+    container.start();
+
+    return transfer;
+  }
+
+  public void sendMessage(String queueName, Object message) {
```
Comment (NickGrantONS) [inline src/test/java/uk/gov/ons/census/notifyprocessor/utilities/RabbitQueueHelper.java]: unused method

File: src/test/java/uk/gov/ons/census/notifyprocessor/utilities/RabbitQueueHelper.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+  @Retryable(
+      value = {IOException.class},
+      maxAttempts = 10,
+      backoff = @Backoff(delay = 5000))
+  public void purgeQueue(String queueName) {
+    amqpAdmin.purgeQueue(queueName);
+  }
+
+  public String getMessage(BlockingQueue<String> queue) throws InterruptedException {
```
Comment (NickGrantONS) [inline src/test/java/uk/gov/ons/census/notifyprocessor/utilities/RabbitQueueHelper.java]: unused method

File: src/test/java/uk/gov/ons/census/notifyprocessor/utilities/RabbitQueueHelper.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    return queue.poll(20, TimeUnit.SECONDS);
+  }
+
+  public void checkNoMessage(BlockingQueue<String> queue) throws InterruptedException {
```
Comment (NickGrantONS) [inline src/test/java/uk/gov/ons/census/notifyprocessor/utilities/RabbitQueueHelper.java]: unused method

File: src/main/java/uk/gov/ons/census/notifyprocessor/client/CaseClient.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+  @Value("${caseapi.port}")
+  private String port;
+
+  public CaseClient(RestTemplate restTemplate) {
+    this.restTemplate = restTemplate;
+  }
+
+  public UacQidDTO getUacQid(String caseId, int questionnaireType) {
+    String url = "http://" + host + ":" + port + "/uacqid/create";
```
Comment (NickGrantONS) [inline src/main/java/uk/gov/ons/census/notifyprocessor/client/CaseClient.java]: I guess you could use `UriComponentsBuilder` here

Comment (hbrace) [reply src/main/java/uk/gov/ons/census/notifyprocessor/client/CaseClient.java]: Done 👍

File: src/main/java/uk/gov/ons/census/notifyprocessor/client/CaseClient.java [hunk 7]
```diff
@@ -22,12 +23,17 @@ public CaseClient(RestTemplate restTemplate) {
...
+            .scheme("http")
+            .host(host)
+            .port(port)
+            .path("/uacqid/create");
     CaseDetailsDTO caseDetails = new CaseDetailsDTO();
     caseDetails.setCaseId(caseId);
     caseDetails.setQuestionnaireType(Integer.toString(questionnaireType));
 
-    UacQidDTO uacQid = restTemplate.postForObject(url, caseDetails, UacQidDTO.class);
+    UacQidDTO uacQid = restTemplate.postForObject(url.toUriString(), caseDetails, UacQidDTO.class);
```
Comment (NickGrantONS) [inline src/main/java/uk/gov/ons/census/notifyprocessor/client/CaseClient.java]: you can just do `uriComponents.toUri()` and not pass in a string

Comment (hbrace) [reply src/main/java/uk/gov/ons/census/notifyprocessor/client/CaseClient.java]: Code updated

File: src/main/resources/application.yml [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+healthcheck:
+  frequency: 1000 #milliseconds
+  filename: /tmp/notify-processor-healthy
+
+notify:
+  apiKey: dummykey-ffffffff-ffff-ffff-ffff-ffffffffffff-ffffffff-ffff-ffff-ffff-ffffffffffff
+  baseUrl: https://dummy-notify:123
+  testMode: true
+  templateIdReplacementHHE: 21447bc2-e7c7- 41ba-8c5e- 7a5893068525
+  templateIdReplacementHHW: ef045f43-ffa8-4047- b8e2-65bfbce0f026
```
Comment (NickGrantONS) [inline src/main/resources/application.yml]: space

**Pull request merged**